### PR TITLE
Fix/labeling service

### PR DESCRIPTION
**Pull request by justjuju-me**
# Description

Fix labeling service

### CODE REVIEW

File: app/services/labeling/relabel_service.rb [hunk 1]
```diff
@@ -2,31 +2,73 @@ module Labeling
...
+      return first_donation_or_contribution_date if DonationContribution.last.nil? && ContributionFee.last.nil?
+      return DonationContribution.last&.donation&.created_at if ContributionFee.last.nil?
+      return ContributionFee.last&.contribution&.created_at if DonationContribution.last.nil?
+
+      [DonationContribution.last&.donation&.next&.created_at,
+       ContributionFee.last&.contribution&.next&.created_at].max
+    end
+
+    def label_unlabelled_records
+      from = next_unlabelled_date
```
Comment (thaisribon) [inline app/services/labeling/relabel_service.rb:52]: i think here should be @from, right?

File: app/services/labeling/relabel_service.rb [hunk 2]
```diff
@@ -2,31 +2,73 @@ module Labeling
...
+            donation = Donation.find(record['id'])
+            deal_with_donation(donation)
+          when 'PersonBlockchainTransaction'
+            contribution = PersonBlockchainTransaction.find(record['id']).person_payment&.contribution
+            deal_with_contribution(contribution) if contribution
+          end
+        end
+
+        @from = next_unlabelled_date
+        break unless next_unlabelled_date.nil?
```
Comment (thaisribon) [inline app/services/labeling/relabel_service.rb:71]: should not be `break if from.nil?` ?

File: app/services/labeling/relabel_service.rb [hunk 3]
```diff
@@ -2,31 +2,73 @@ module Labeling
...
+      [Donation.first&.created_at, Contribution.first&.created_at].min
+    end
+
+    # rubocop:disble Metrics/AbcSize
+    def next_unlabelled_date
+      return first_donation_or_contribution_date if DonationContribution.last.nil? && ContributionFee.last.nil?
+      return DonationContribution.last&.donation&.created_at if ContributionFee.last.nil?
+      return ContributionFee.last&.contribution&.created_at if DonationContribution.last.nil?
+
+      [DonationContribution.last&.donation&.next&.created_at,
```
Comment (thaisribon) [inline app/services/labeling/relabel_service.rb:47]: what is next?

File: app/services/labeling/relabel_service.rb [hunk 4]
```diff
@@ -2,31 +2,73 @@ module Labeling
...
+        @contributions = Contribution.where('created_at >= ?', from)
+        ContributionBalance.where(contribution: @contributions).delete_all
+        @contributions.each(&:set_contribution_balance)
+        ContributionFee.where(contribution: @contributions).delete_all
+
+        DonationContribution.where(donation: Donation.where('created_at >= ?', from)).delete_all
+      end
+    end
+
+    def last_donation_or_contribution_date
```
Comment (thaisribon) [inline app/services/labeling/relabel_service.rb:25]: i think you are not using this method, right?

**Review by thaisribon — APPROVED** [review]

File: app/services/labeling/relabel_service.rb (hunk 3 above)
Comment (justjuju-me) [reply app/services/labeling/relabel_service.rb:47]: "next" get the next donation id. In this case, i will get the last labelled donation, then get the next donation id, that should be the next record to label.

**Pull request closed without merging**